### PR TITLE
Mm 07 navbar

### PR DESCRIPTION
**Pull request by tomscambler**
We have added a navbar which appears when you click on the hamburger in the top left corner. It disappears when you click on the cross.

### CODE REVIEW

**Comment by TimLeach635** [comment]
image
Screenshot demonstration

**Comment by tomscambler** [comment]
image

**Comment by tomscambler** [comment]
image

File: src/components/navigation/Navbar.tsx [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
 
 export const Navbar: React.FunctionComponent = () => {
   return <nav>
-    <ul>
-      <li><Link to="/">Home</Link></li>
-      <li><Link to="/rovers">Rovers</Link></li>
-      <li><Link to="/quiz">Quiz</Link></li>
-    </ul>
+
+    <div className="emptySpacingBox"></div>
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: Having empty elements just to add space isn't very accessible. Could you achieve the same effect by using a top margin, or setting the position of the element directly?

File: src/components/navigation/Navbar.tsx [hunk 2]
```diff
@@ -1,12 +1,25 @@
...
+
+    <div id="menuToggle">
+      <input type="checkbox" />
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: This is very clever! You're using [the checkbox hack](https://css-tricks.com/the-checkbox-hack/), which is some seriously advanced stuff. Well done!

However, from that article:
image

"Questionable semantics" is right - this solution:
1. is fairly brittle - it would only take a small bug/difference in CSS handling to totally wreck it
2. isn't semantic or accessible - a screen reader or other HTML-focused viewers would have hard times figuring out what was going on with this

Functionality/interactivity should (as the article suggests) be implemented using Javascript. React makes this very easy - you can store the boolean "is the menu open" variable in the _state_ of the navbar component, and have a class appear conditionally on your elements:
```
<nav className={`menu ${isMenuOpen ? "expanded" : "collapsed"}`}>
```
Then use a `<button>` (it is probably semantically the element that makes most sense here) to toggle that state!

So, in summary - very clever solution indeed, but I'm sorry to say not appropriate for an accessible website!

File: src/components/navigation/Navbar.tsx [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
+      
+        <span></span>
+        <span></span>
+        <span></span>
+      
+        <ul id="menu">
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: Prefer using classes over IDs:
```suggestion
        <ul className="menu">
```

and also, "menu" is a little generic - there's this neat CSS class naming convention called [BEM](https://codeburst.io/understanding-css-bem-naming-convention-a8cca116d252) that has some good ideas. I think here we would say the block is the navbar, and the element is the list of links, so this class may be called `navbar__nav-links`!

File: src/components/navigation/Navbar.scss [hunk 4]
```diff
@@ -0,0 +1,148 @@
+@import '../../constants.scss';
+
+.emptySpacingBox{
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Prefer space before brace
```suggestion
.emptySpacingBox {
```

File: src/components/navigation/Navbar.scss [hunk 5]
```diff
@@ -0,0 +1,148 @@
+@import '../../constants.scss';
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Prefer double quotes (`"`)
```suggestion
@import "../../constants.scss";
```

File: src/components/navigation/Navbar.scss [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+.emptySpacingBox{
+  width:100%;
+  height:50px;
+}
+
+#menuToggle
+{
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Prefer brace on same line for CSS (please change the others in this file too!)
```suggestion
#menuToggle {
```

File: src/components/navigation/Navbar.scss [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  -webkit-user-select: none;
+  user-select: none;
+}
+
+#menuToggle a
+{
+  text-decoration: none;
+  color: white;
+  transition: color 0.3s, underline 0.3s;
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: We're using SCSS, so all of these can be grouped/nested for clarity!
```suggestion
  user-select: none;
  
  a  {
    text-decoration: none;
    color: white;
    transition: color 0.3s, underline 0.3s;
  }
}
```

File: src/components/navigation/Navbar.scss [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  transform: rotate(0deg) scale(0.2, 0.2);
+}
+
+/*
+ * Ohyeah and the last one should go the other direction
+ */
+#menuToggle input:checked ~ span:nth-last-child(2)
+{
+  transform: rotate(-45deg) translate(0, -1px);
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: All of these can be brought inside the `#menuToggle` definition as in the above comment

File: src/components/navigation/Navbar.tsx [hunk 9]
```diff
@@ -1,12 +1,25 @@
...
   return <nav>
-    <ul>
-      <li><Link to="/">Home</Link></li>
-      <li><Link to="/rovers">Rovers</Link></li>
-      <li><Link to="/quiz">Quiz</Link></li>
-    </ul>
+
+    <div className="emptySpacingBox"></div>
+
+    <div id="menuToggle">
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: Prefer classes to ids
```suggestion
    <div className="menuToggle">
```

File: src/components/navigation/Navbar.scss [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+/*
+ * Make this absolute positioned
+ * at the top left of the screen
+ */
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss:66]: Have you thought about other screen sizes here, and if so, is this your choice for all screen sizes, or just mobile?

File: src/components/navigation/Navbar.tsx [hunk 11]
```diff
@@ -1,12 +1,25 @@
...
+      <input type="checkbox" />
+      
+        <span></span>
+        <span></span>
+        <span></span>
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: So... this is a very common way of doing this, which is why I'm not going to ask you not to do it, but hopefully you can see that having empty HTML elements purely for visuals isn't very semantic/accessible! Don't change this at this point, but I think the most accessible way of doing this is to use an SVG icon, possibly animating it somehow.

(You're probably noticing at this point that I value accessibility over most other things! 😉)

File: src/constants.scss [hunk 12]
```diff
@@ -3,3 +3,15 @@ $background-light-accent: #bbbbbb;
...
+$grey-blue: #8ba6a9;
+$light-blue: #a7cecb;
+$dark-orange: #cf5c36;
+$light-orange: #f1a66a;
+
+$primary: $dark-orange;
+$secondary: $dark-blue;
+$primary-light: $light-orange;
+$secondary-light: $light-blue;
+
```
Comment (TimLeach635) [inline src/constants.scss]: Exactly one blank line at the end of the file please!

File: src/components/navigation/Navbar.scss [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  transform: translate(-100%, 0);
+  
+  transition: transform 0.5s cubic-bezier(0.77,0.2,0.05,1.0);
+}
+
+#menu li
+{
+  padding: 10px 0;
+  font-size: 22px;
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Can be brought up inside the `#menu` definition

**Review by TimLeach635 — CHANGES_REQUESTED** [review]
A few things to change, really really clever stuff in here!

File: src/components/navigation/Navbar.scss (hunk 10 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.scss:66]: Yes this is our choice for all sizes - it looks ok on larger screens too

File: src/components/navigation/Navbar.tsx (hunk 1 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.tsx]: Have left this out for individual pages to deal with

File: src/components/navigation/Navbar.tsx (hunk 2 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.tsx]: Done

File: src/components/navigation/Navbar.tsx (hunk 3 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.tsx]: Done

File: src/components/navigation/Navbar.scss (hunk 4 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.scss]: Done

File: src/components/navigation/Navbar.scss (hunk 5 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.scss]: Done

File: src/components/navigation/Navbar.tsx (hunk 9 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.tsx]: Done

File: src/components/navigation/Navbar.tsx (hunk 11 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.tsx]: Noted

File: src/components/navigation/Navbar.scss (hunk 13 above)
Comment (rosalievan) [reply src/components/navigation/Navbar.scss]: Done

File: src/components/navigation/Navbar.scss [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+.navbar a {
+  text-decoration: none;
+  color: white;
+  transition: color 0.3s, underline 0.3s;
+}
+
+.navbar a:hover {
+  color: black;
+  text-decoration: underline;
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: These can be brought inside the `.navbar` thanks to SCSS
```suggestion

  a {
    text-decoration: none;
    color: white;
    transition: color 0.3s, underline 0.3s;

    $:hover {
      color: black;
      text-decoration: underline;
    }
  }
}
```

File: src/components/navigation/Navbar.scss [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+              opacity 0.55s ease;
+}
+
+.navbar span:first-child {
+  transform-origin: 0% 0%;
+}
+
+.navbar span:nth-last-child(2) {
+  transform-origin: 0% 100%;
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Likewise above can be nested

File: src/components/navigation/Navbar.scss [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+.burgerexpanded {
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: In BEM, "expanded" is a modifier (the "M" in BEM) so we'd probably call this class:
`.navbar__burger--expanded`

File: src/components/navigation/Navbar.scss [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+  &.navbar__spanone {
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Prefer `kebab-casing` for classes, so `.navbar__span-one`

File: src/components/navigation/Navbar.scss [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    transform: rotate(45deg) translate(0, -3px);
+    background-color: white;
+  }
+
+  &.navbar__spantwo {
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: `span-two`

File: src/components/navigation/Navbar.scss [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    opacity: 0;
+    background-color: white;
+  }
+
+  &.navbar__spanthree {
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: `span-three`

File: src/components/navigation/Navbar.scss (hunk 10 above)
Comment (TimLeach635) [reply src/components/navigation/Navbar.scss:66]: 👍🏻

File: src/components/navigation/Navbar.scss [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+ */
+.expanded {
+  transform: none;
+}
+
+.navbar button {
+  background: transparent;
+  border: transparent;
+  z-index: 4 !important;
+}
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.scss]: Nest this inside `.navbar` as above

File: src/components/navigation/Navbar.tsx [hunk 21]
```diff
@@ -1,12 +1,33 @@
...
-      <li><Link to="/rovers">Rovers</Link></li>
-      <li><Link to="/quiz">Quiz</Link></li>
-    </ul>
+
+    <div className="navbar">
+      <button onClick = {() => toggleMenu()}>
+      
+        <span className = {`navbar__spanone ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
+        <span className = {`navbar__spantwo ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
+        <span className = {`navbar__spanthree ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: Prefer `kebab-case`, so `burger-expanded` and `burger-collapsed` (but you may be able to make it cleaner by either just using `expanded` and `collapsed`, or full BEM with `navbar__hamburger-line--expanded` etc)

File: src/components/navigation/Navbar.tsx [hunk 22]
```diff
@@ -1,12 +1,33 @@
...
+
+      </button>
+
```
Comment (TimLeach635) [inline src/components/navigation/Navbar.tsx]: A few unnecessary blank lines:
```suggestion
    <div className="navbar">
      <button onClick = {() => toggleMenu()}>
        <span className = {`navbar__spanone ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
        <span className = {`navbar__spantwo ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
        <span className = {`navbar__spanthree ${isExpanded ? "burgerexpanded" : "burgercollapsed"}`}></span>
      </button>
```

**Review by TimLeach635 — CHANGES_REQUESTED** [review]
Looks great, very nearly done!

File: src/components/navigation/Navbar.scss (hunk 14 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.scss]: Done - did you mean &:hover on 18?

File: src/components/navigation/Navbar.scss (hunk 15 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.scss]: done, thanks

File: src/components/navigation/Navbar.scss (hunk 16 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.scss]: done, thanks

File: src/components/navigation/Navbar.scss (hunk 17 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.scss]: done, thanks

File: src/components/navigation/Navbar.tsx (hunk 21 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.tsx]: we've gone for navbar__nav-links--expanded

File: src/components/navigation/Navbar.tsx (hunk 22 above)
Comment (tomscambler) [reply src/components/navigation/Navbar.tsx]: done, thanks

**Review by TimLeach635 — APPROVED** [review]
Very very well done! Really nicely crafted component, you should be proud
